DAOS-16686 dfuse: Fix overlapping chunk reads.

ashleypittman opened the pull request:
Handle concurrent read in the chunk_read code.  Rather than assuming
each slot only gets requested once save the slot number as part of the
request and handle multiple requests.

This corrects the behaviour and avoids a crash when multiple readers read
the same file concurrently and improves the performance in this case.

ashleypittman committed:
DAOS-15682 dfuse: Fail on concurrent read.

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

ashleypittman committed:
Try and fix issue.

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

ashleypittman committed:
First stab at a fix.

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

ashleypittman committed:
Merge branch 'master' into amd/dfuse-concurrent-read

ashleypittman committed:
Fix invalid free and leak.

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

ashleypittman committed:
Fix a logging line.

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

ashleypittman committed:
Add some debugging.

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

daosbuild1 commented:
Test stage Functional Hardware Large completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-15298/4/testReport/

ashleypittman committed:
Track duplicate reads.  This avoids a crash but there's still a memory leak

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

ashleypittman committed:
Fix logic.

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

ashleypittman committed:
Rework to support blocking on network requests.

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

ashleypittman committed:
Bump array size and add stats.

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

daosbuild1 commented:
Test stage Build RPM on EL 9 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15298/6/execution/node/357/log

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15298/6/execution/node/356/log

daosbuild1 commented:
Test stage Build RPM on Leap 15.5 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15298/6/execution/node/351/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15298/6/execution/node/348/log

ashleypittman committed:
Fix a segv in the stats.

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

ashleypittman committed:
Track EOF better in reads.

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

daosbuild1 commented:
Test stage NLT on EL 8.8 completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-15298/8/testReport/

ashleypittman committed:
Merge branch 'master' into amd/dfuse-concurrent-read

ashleypittman committed:
Fixup after merge

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

ashleypittman committed:
Move active read list to active.

Test-tag: dfuse

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

daosbuild1 commented:
Test stage NLT on EL 8.8 completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-15298/9/testReport/

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15298/9/execution/node/1479/log

ashleypittman committed:
Merge branch 'master' into amd/dfuse-concurrent-read

Test-tag: dfuse

ashleypittman committed:
Rebase and iterate on comments.

Test-tag: dfuse
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

ashleypittman committed:
fix: remove an extra list operation.

Test-tag: dfuse
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

daosbuild1 commented:
Test stage NLT on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-15298/11/display/redirect

daosbuild1 commented:
Test stage NLT on EL 8.8 completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-15298/12/testReport/

ashleypittman commented on src/client/dfuse/dfuse.h:
```diff
@@ -474,6 +481,10 @@ struct dfuse_pool {
 	ACTION(OPEN)                                                                               \
 	ACTION(PRE_READ)                                                                           \
 	ACTION(READ)                                                                               \
+	ACTION(READ_EOF_M)                                                                         \
+	ACTION(READ_CON)                                                                           \
+	ACTION(READ_BUCKET)                                                                        \
+	ACTION(READ_BUCKET_M)                                                                      \
```
Extra metrics in dfuse is a good thing but I'm thinking of removing these from this PR.  They were certainly useful to detect/different code-paths but I think I'd rather extend these outside of a bug-fix PR.

phender commented on src/tests/ftest/dfuse/read.py:
```diff
@@ -23,7 +23,9 @@ def test_dfuse_pre_read(self):
         the first file leaves the flag enabled.
 
-        :avocado: tags=all,full_regression
+        TODO: Check this with 128k I/O size and non-128K I/O size.
```
I'm surprised this isn't be called out by the linting check.  We should either do this or remove the comment before landing and add a ticket as a reminder.

phender commented on src/tests/ftest/dfuse/read.py:
```diff
@@ -23,7 +23,9 @@ def test_dfuse_pre_read(self):
-        :avocado: tags=all,full_regression
+        TODO: Check this with 128k I/O size and non-128K I/O size.
+
+        :avocado: tags=all,daily_regression
```
Just confirming that moving this from weekly to daily is intentional (I thought I had asked this before, but couldn't find the question in the ticket).  According to the [last weekly](https://build.hpdd.intel.com/job/daos-stack/job/daos/job/weekly-testing/411/testReport/FTEST_dfuse/DFusePreReadTest/) it only takes a minute to run, so the move should be fine.

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15298/12/execution/node/1480/log

ashleypittman committed:
Merge branch 'master' into amd/dfuse-concurrent-read

ashleypittman committed:
Back out test and stat changes.

Test-tag: dfuse

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

daosbuild1 commented:
Test stage NLT on EL 8.8 completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-15298/13/testReport/

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15298/13/execution/node/1463/log

ashleypittman committed:
Merge branch 'master' into amd/dfuse-concurrent-read

ashleypittman committed:
Try and solve patchelf problem.

Test-tag: DaosBuild
Skip-fault-injection-test: true

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

daosbuild1 commented:
Test stage Build RPM on Leap 15.5 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15298/14/execution/node/328/log

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15298/14/execution/node/384/log

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15298/15/execution/node/1372/log

ashleypittman commented:
This is consistently failing one of the tests, taking back to draft whilst I investigate.

ashleypittman committed:
Change failure mode.

Skip-func-test-vm: true

Test-tag: DaosBuild
Skip-fault-injection-test: true
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

ashleypittman committed:
Merge branch 'master' into amd/dfuse-concurrent-read

Test-tag: dfuse

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

daosbuild1 commented:
Test stage Functional on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15298/17/execution/node/1201/log

ashleypittman committed:
Merge branch 'master' into amd/dfuse-concurrent-read

Test-tag: dfuse

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15298/18/execution/node/1428/log

ashleypittman committed:
add strace.

Skip-func-test-vm: true

Test-tag: DaosBuild
Skip-fault-injection-test: true

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

ashleypittman committed:
Do not pre-fetch or keep cache for writeable files.

Skip-unit-tests: true
Skip-func-test-vm: true
Test-tag: DaosBuild
Skip-fault-injection-test: true
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

ashleypittman committed:
Stat and re-fresh the files.

Skip-unit-tests: true
Skip-func-test-vm: true
Test-tag: DaosBuild
Skip-fault-injection-test: true
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

ashleypittman committed:
Fix build not to fail.

Skip-unit-tests: true
Skip-func-test-vm: true
Test-tag: DaosBuild
Skip-fault-injection-test: true
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15298/25/execution/node/936/log

ashleypittman committed:
Stat the file, not the link

Skip-unit-tests: true
Skip-func-test-vm: true
Test-tag: DaosBuild
Skip-fault-injection-test: true
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15298/26/execution/node/936/log

ashleypittman committed:
Disable matching reads and chunk code to debug failure.

Skip-unit-tests: true
Skip-func-test-vm: true
Test-tag: DaosBuild
Skip-fault-injection-test: true
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15298/27/execution/node/936/log

ashleypittman committed:
Disalbe the file-size EOF optimisation.

Skip-unit-tests: true
Skip-func-test-vm: true
Test-tag: DaosBuild
Skip-fault-injection-test: true
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

ashleypittman committed:
Re-enable feature.

Skip-unit-tests: true
Skip-func-test-vm: true
Test-tag: DaosBuild
Skip-fault-injection-test: true
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

ashleypittman committed:
Back out build debug.

Skip-unit-tests: true
Skip-func-test-vm: true
Test-tag: DaosBuild
Skip-fault-injection-test: true
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

ashleypittman committed:
Merge branch 'master' into amd/dfuse-concurrent-read

ashleypittman committed:
Back out matching read changes, they're in a different PR.

Skip-unit-tests: true
Skip-func-test-vm: true
Test-tag: DaosBuild
Skip-fault-injection-test: true

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

ashleypittman committed:
Make number of slots vcariable.

Skip-unit-tests: true
Skip-func-test-vm: true
Test-tag: DaosBuild
Skip-fault-injection-test: true
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

daosbuild1 commented:
Test stage Build on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15298/31/execution/node/388/log

daosbuild1 commented:
Test stage Build on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15298/32/execution/node/388/log

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15298/32/execution/node/361/log

daosbuild1 commented:
Test stage Build RPM on Leap 15.5 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15298/32/execution/node/359/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15298/32/execution/node/360/log

ashleypittman committed:
Fix a compile warning.

Skip-unit-tests: true
Skip-func-test-vm: true
Test-tag: DaosBuild
Skip-fault-injection-test: true
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

ashleypittman committed:
Handle truncated reads and avoid crash.

Skip-unit-tests: true
Skip-func-test-vm: true
Test-tag: DaosBuild
Skip-fault-injection-test: true
Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

ashleypittman commented on src/client/dfuse/ops/read.c:
```diff
@@ -173,186 +173,112 @@ pick_eqt(struct dfuse_info *dfuse_info)
+	D_SPIN_UNLOCK(&ia->lock);
 
-		D_SPIN_UNLOCK(&ia->lock);
+	d_list_for_each_entry_safe(cr, crn, &cd->req_list, req_list) {
```
Note to self: There's no reference on `cd` held here after the last call to `DFUSE_REPLY../.` so the list needs to be spliced onto the stack before the list is iterated.

jolivier23 committed:
Merge branch 'master' into amd/dfuse-concurrent-read

wangdi1 committed:
DAOS-16686 dfuse: avoid possible corruption

Comments from Ashely

In chunk_cb(), there's no reference on cd held here after the
last call to DFUSE_REPLY../. so the list needs to be spliced
onto the stack before the list is iterated.

Required-githooks: true

Signed-off-by: Di Wang <ddiwang@google.com>

wangdi1 committed:
Merge branch 'master' into amd/dfuse-concurrent-read

phender reviewed (dismissed):
Removing -1.

mchaarawi reviewed (changes requested):
change looks ok to me. but since the description mentions an issue with a crash and not just a perf improvement, a test needs to be added.

Also please push with Features: dfuse

ashleypittman commented:
> change looks ok to me. but since the description mentions an issue with a crash and not just a perf improvement, a test needs to be added.

I actually wrote a PR to test this issue, it's easy enough to trigger locally but the timing in CI meant that I wasn't able to reproduce it via ftest, possibly because reads are just faster than running over loopback.

https://github.com/daos-stack/daos/pull/15312

ashleypittman commented:
> Yea i see how this can be a race that is not always reproducible. but at least having that test in would still give someone a way to run that.

> curious though if putting that case in NLT would trigger that issue since NLT runs on 1 host?

Yes, it's reproducible in NLT every time.  In that case dfuse is run under valgrind so will be much slower and hence it's reproducible every time.  I didn't want to do that though as NLT runs with full debug on so I've typically only used that for very small datasets/iteration counts.  I don't see another way to get this into CI though.

I'll paste the script I've been using tomorrow when I'm at my work laptop but it's the same script I copied to make that test, essentially create a 1M file in dfuse, evict it and then read it in parallel in at least 128k chunks.  The bug/race of course is when the same 128k bunk of a file is requested multiple times concurrently before the first request has been replied to.

ashleypittman commented:
I can't seem to push to this PR today for some reason but it want's [c7a870377d632454fd99c97dc7cdb4042b01d444](https://github.com/daos-stack/daos/pull/15680/commits/c7a870377d632454fd99c97dc7cdb4042b01d444) added to it.

jolivier23 committed:
Merge branch 'master' into amd/dfuse-concurrent-read

jolivier23 committed:
Patch from Ashley

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

daosbuild1 commented:
Test stage Build RPM on EL 9 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15298/37/execution/node/342/log

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15298/37/execution/node/339/log

daosbuild1 commented:
Test stage Build RPM on Leap 15.5 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15298/37/execution/node/345/log

jolivier23 committed:
Test-tag: dfuse
Allow-unstable-test: true

Signed-off-by: Jeff Olivier <jeffolivier@google.com>

daosbuild1 commented:
Test stage Build RPM on EL 9 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15298/38/execution/node/338/log

daosbuild1 commented:
Test stage Build on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15298/38/execution/node/361/log

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15298/38/execution/node/357/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15298/38/execution/node/341/log

daosbuild1 commented:
Test stage Build on Leap 15.5 with Intel-C and TARGET_PREFIX completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-15298/38/execution/node/481/log

jolivier23 committed:
Merge branch 'master' into amd/dfuse-concurrent-read

jolivier23 committed:
Fix a typo

Allow-unstable-test: true

Signed-off-by: Jeff Olivier <jeffolivier@google.com>

wangdi1 committed:
Merge branch 'master' into amd/dfuse-concurrent-read

Features: dfuse

wangdi1 committed:
Merge branch 'master' into amd/dfuse-concurrent-read

wangdi1 committed:
DAOS-16686 dfuse: fix copywrite

fix copywrite

Features: dfuse
Signed-off-by: Di Wang <ddiwang@google.com>

wangdi1 committed:
Features: dfuse
Merge branch 'master' into amd/dfuse-concurrent-read

Commit:
Merge branch 'master' into amd/dfuse-concurrent-read

Signed-off-by: Johann Lombardi <johann.lombardi@hpe.com>

mchaarawi committed:
Merge branch 'master' into amd/dfuse-concurrent-read

daosbuild3 commented:
Test stage Functional Hardware Medium MD on SSD completed with status UNSTABLE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos//view/change-requests/job/PR-15298/7/testReport/

daosbuild3 commented:
Test stage Functional Hardware Medium MD on SSD completed with status UNSTABLE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos//view/change-requests/job/PR-15298/8/testReport/

daosbuild3 commented:
Test stage Functional Hardware Medium MD on SSD completed with status UNSTABLE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos//view/change-requests/job/PR-15298/9/testReport/
